### PR TITLE
[#37] Gitlab Connector: Add activity part

### PR DESCRIPTION
**Pull request by BeckerFrank**
* Separate Comments from other System Activity Events.
* use StyledText to show the Activity Events with color Background
* add Tooltip and open Action to the Design Events

### CODE REVIEW

File: mylyn.tasks/connectors/gitlab/org.eclipse.mylyn.gitlab.core/src/org/eclipse/mylyn/internal/gitlab/core/GitlabRestClient.java [hunk 1]
```diff
@@ -490,39 +503,296 @@ public TaskData getTaskData(TaskRepository repository, String taskId, IProgressM
...
 			}
 		    }
 		}
 	    }
 
+	    JsonArray states = getIssueStateEvents(matcher.group(1), matcher.group(2), OperationUtil.convert(monitor));
+	    if (states != null) {
+		for (JsonElement stateElem : states) {
+		    JsonObject state = (JsonObject) stateElem;
+		    Date date = GitlabTaskAttributeMapper.parseDate(state.get("created_at").getAsString());
```
Comment (gnl42) [inline mylyn.tasks/connectors/gitlab/org.eclipse.mylyn.gitlab.core/src/org/eclipse/mylyn/internal/gitlab/core/GitlabRestClient.java:553]: Interval?

File: mylyn.tasks/connectors/gitlab/org.eclipse.mylyn.gitlab.core/src/org/eclipse/mylyn/internal/gitlab/core/GitlabRestClient.java [hunk 2]
```diff
@@ -468,6 +473,14 @@ public String obtainAccessToken(IOperationMonitor monitor) throws Exception {
     public TaskData getTaskData(TaskRepository repository, String taskId, IProgressMonitor monitor)
 	    throws GitlabException {
 	TaskData result = null;
+	long startTime, endTime = 0;
+	Thread thread;
+	if (TRACE) {
+	    thread = Thread.currentThread();
+	    System.out.println(
+		    thread.getName() + " getTaskData start: " + repository.getRepositoryUrl() + " id " + taskId);
+	    startTime = Calendar.getInstance().getTimeInMillis();
```
Comment (gnl42) [inline mylyn.tasks/connectors/gitlab/org.eclipse.mylyn.gitlab.core/src/org/eclipse/mylyn/internal/gitlab/core/GitlabRestClient.java:482]: `System.currentTimeMillis()`

File: mylyn.tasks/connectors/gitlab/org.eclipse.mylyn.gitlab.core/src/org/eclipse/mylyn/internal/gitlab/core/GitlabRepositoryConnector.java [hunk 3]
```diff
@@ -162,12 +163,28 @@ public Optional<GitlabConfiguration> load(RepositoryKey key) throws Exception {
...
     public GitlabConfiguration getRepositoryConfiguration(TaskRepository repository) throws CoreException {
+	long startTime, endTime = 0;
+	Thread thread;
 	if (clientCache.getIfPresent(new RepositoryKey(repository)) == null) {
 	    getClient(repository);
 	}
 	try {
+	    if (TRACE) {
+		thread = Thread.currentThread();
+		startTime = Calendar.getInstance().getTimeInMillis();
```
Comment (gnl42) [inline mylyn.tasks/connectors/gitlab/org.eclipse.mylyn.gitlab.core/src/org/eclipse/mylyn/internal/gitlab/core/GitlabRepositoryConnector.java:174]: `System.currentTimeMillis()`

File: mylyn.tasks/connectors/gitlab/org.eclipse.mylyn.gitlab.core/src/org/eclipse/mylyn/internal/gitlab/core/GitlabRestClient.java [hunk 4]
```diff
@@ -490,39 +503,296 @@ public TaskData getTaskData(TaskRepository repository, String taskId, IProgressM
...
 	return result;
     }
 
+    private void buildLableText(TaskAttribute labelText, ArrayList<String> added, ArrayList<String> removed) {
+	if (labelText != null) {
+	    String text = "";
+	    if (added.size() > 0) {
+		text += "added ";
+		text += String.join(", ", added);
+		if (added.size() > 1) {
```
Comment (gnl42) [inline mylyn.tasks/connectors/gitlab/org.eclipse.mylyn.gitlab.core/src/org/eclipse/mylyn/internal/gitlab/core/GitlabRestClient.java:647]: `text += added.size() == 1 ? "label" : "labels";`

File: mylyn.tasks/connectors/gitlab/org.eclipse.mylyn.gitlab.core/src/org/eclipse/mylyn/internal/gitlab/core/GitlabRestClient.java [hunk 5]
```diff
@@ -490,39 +503,296 @@ public TaskData getTaskData(TaskRepository repository, String taskId, IProgressM
...
+		    text += " labels";
+		} else {
+		    text += " label";
+		}
+	    }
+	    if (removed.size() > 0) {
+		if (text.length() == 0) {
```
Comment (gnl42) [inline mylyn.tasks/connectors/gitlab/org.eclipse.mylyn.gitlab.core/src/org/eclipse/mylyn/internal/gitlab/core/GitlabRestClient.java:654]: `text += text.length() == 0 ? "removed" : "and removed"`

File: mylyn.tasks/connectors/gitlab/org.eclipse.mylyn.gitlab.core/src/org/eclipse/mylyn/internal/gitlab/core/GitlabRestClient.java [hunk 6]
```diff
@@ -490,39 +503,296 @@ public TaskData getTaskData(TaskRepository repository, String taskId, IProgressM
...
+	    typeAttribute.setValue(ActivityType.LOCK.toString());
+	} else if (bodyText.contains("/designs?version=")) {
+	    typeAttribute.setValue(ActivityType.DESIGN.toString());
+	} else {
+	    typeAttribute.setValue(ActivityType.UNKNOWN.toString());
+	}
+
+	TaskAttribute styleAttribute = taskAttribute.createAttribute(GitlabCoreActivator.ATTRIBUTE_TYPE_ACTIVITY_STYLE);
+	String resultText = "";
+	Pattern r = Pattern.compile("\\[(.+)\\]\\((.+)\\)");
```
Comment (gnl42) [inline mylyn.tasks/connectors/gitlab/org.eclipse.mylyn.gitlab.core/src/org/eclipse/mylyn/internal/gitlab/core/GitlabRestClient.java:711]: `private static final `

File: mylyn.tasks/connectors/gitlab/org.eclipse.mylyn.gitlab.ui/src/org/eclipse/mylyn/internal/gitlab/ui/TaskEditorActivityPart.java [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+    }
+
+    private void createActivityList(FormToolkit toolkit) {
+	for (TaskAttribute taskAttribute : activityAttributes) {
+	    String activityText = taskAttribute.getAttribute(TaskAttribute.COMMENT_TEXT).getValue();
+	    String activityType = taskAttribute.getAttribute(GitlabCoreActivator.ATTRIBUTE_TYPE_ACTIVITY).getValue();
+
+	    Label labelIcon = toolkit.createLabel(activityComposite, "");
+
+	    switch (GitlabCoreActivator.ActivityType.valueOf(activityType)) {
```
Comment (gnl42) [inline mylyn.tasks/connectors/gitlab/org.eclipse.mylyn.gitlab.ui/src/org/eclipse/mylyn/internal/gitlab/ui/TaskEditorActivityPart.java:132]: Undecided if something like:

```
labelIcon.setImage(
			GitlabUiActivator.getDefault().getImageRegistry().get(
			    switch(GitlabCoreActivator.ActivityType.valueOf(activityType)) {
			        CALENDAR: yield GitlabUiActivator.GITLAB_CALENDAR_FILE
			    ...
                          }
                        )
	);
```
would be easier?

File: mylyn.tasks/connectors/gitlab/org.eclipse.mylyn.gitlab.core/src/org/eclipse/mylyn/internal/gitlab/core/GitlabActivityStyle.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+    public void setUrl(String url) {
+	this.url = url;
+    }
+
+    @Override
+    public GitlabActivityStyle clone() throws CloneNotSupportedException {
+	return (GitlabActivityStyle) super.clone();
+    }
+
```
Comment (gnl42) [inline mylyn.tasks/connectors/gitlab/org.eclipse.mylyn.gitlab.core/src/org/eclipse/mylyn/internal/gitlab/core/GitlabActivityStyle.java:87]: add toString()?

I hate debugging code and having to dig into a variable to see it's real value

File: mylyn.tasks/connectors/gitlab/org.eclipse.mylyn.gitlab.core/src/org/eclipse/mylyn/internal/gitlab/core/GitlabRestClient.java (hunk 1 above)
Comment (BeckerFrank) [reply mylyn.tasks/connectors/gitlab/org.eclipse.mylyn.gitlab.core/src/org/eclipse/mylyn/internal/gitlab/core/GitlabRestClient.java]: Done with new commit.

File: mylyn.tasks/connectors/gitlab/org.eclipse.mylyn.gitlab.core/src/org/eclipse/mylyn/internal/gitlab/core/GitlabRestClient.java (hunk 2 above)
Comment (BeckerFrank) [reply mylyn.tasks/connectors/gitlab/org.eclipse.mylyn.gitlab.core/src/org/eclipse/mylyn/internal/gitlab/core/GitlabRestClient.java]: Done with new commit.

File: mylyn.tasks/connectors/gitlab/org.eclipse.mylyn.gitlab.core/src/org/eclipse/mylyn/internal/gitlab/core/GitlabRepositoryConnector.java (hunk 3 above)
Comment (BeckerFrank) [reply mylyn.tasks/connectors/gitlab/org.eclipse.mylyn.gitlab.core/src/org/eclipse/mylyn/internal/gitlab/core/GitlabRepositoryConnector.java]: Done with new commit.

File: mylyn.tasks/connectors/gitlab/org.eclipse.mylyn.gitlab.core/src/org/eclipse/mylyn/internal/gitlab/core/GitlabRestClient.java (hunk 4 above)
Comment (BeckerFrank) [reply mylyn.tasks/connectors/gitlab/org.eclipse.mylyn.gitlab.core/src/org/eclipse/mylyn/internal/gitlab/core/GitlabRestClient.java]: Done with new commit.

File: mylyn.tasks/connectors/gitlab/org.eclipse.mylyn.gitlab.core/src/org/eclipse/mylyn/internal/gitlab/core/GitlabRestClient.java (hunk 5 above)
Comment (BeckerFrank) [reply mylyn.tasks/connectors/gitlab/org.eclipse.mylyn.gitlab.core/src/org/eclipse/mylyn/internal/gitlab/core/GitlabRestClient.java]: Done with new commit.

File: mylyn.tasks/connectors/gitlab/org.eclipse.mylyn.gitlab.core/src/org/eclipse/mylyn/internal/gitlab/core/GitlabRestClient.java (hunk 6 above)
Comment (BeckerFrank) [reply mylyn.tasks/connectors/gitlab/org.eclipse.mylyn.gitlab.core/src/org/eclipse/mylyn/internal/gitlab/core/GitlabRestClient.java]: Done with new commit.

File: mylyn.tasks/connectors/gitlab/org.eclipse.mylyn.gitlab.core/src/org/eclipse/mylyn/internal/gitlab/core/GitlabActivityStyle.java (hunk 8 above)
Comment (BeckerFrank) [reply mylyn.tasks/connectors/gitlab/org.eclipse.mylyn.gitlab.core/src/org/eclipse/mylyn/internal/gitlab/core/GitlabActivityStyle.java:87]: Done with new commit.

**Comment by BeckerFrank** [comment]
@gnl42: can you please look if I fixed all your comments so that we can merge this?

**Review by wimjongman — APPROVED** [review]
Thanks, Frank.

**Pull request merged**